[firebase_crashlytics] runtime exceptions

hffmnn opened the pull request:
## Description

This PR adds support to send non-`FlutterErrorDetails` errors to Crashlytics that have not been caught by the `Flutter.onError` handler.

When e.g. a exception is thrown in a `Future`, then the error handler does not work and the crash will never be reported. By using `runZoned` and using its `onError` handler it is possible to catch those, but the current plugin only supports sending `FlutterErrorDetails` errors. I updated the example to make this point more clear.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. [shared_preferences]
- [x] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [x] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).
- [ ] No, this is *not* a breaking change.

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://www.dartlang.org/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

hffmnn commented on packages/firebase_crashlytics/lib/src/firebase_crashlytics.dart:
```diff
@@ -56,6 +56,34 @@ class Crashlytics {
+      final dynamic result = await channel
+          .invokeMethod<dynamic>('Crashlytics#onError', <String, dynamic>{
+        'exception': "${exception.toString()}",
+        'context': 'onRuntimeException',
```
I am not sure what to send here, because we do not have `ErrorDescription`.

hffmnn replied:
Ok, I double checked: The native method signatures the context gets send to are:
 `- (void)recordCustomExceptionName:(NSString *)name reason:(nullable NSString *)reason frameArray:(CLS_GENERIC_NSARRAY(CLSStackFrame *) *)frameArray;` on iOS. In the end the `context` gets set on the `reason` parameter, which is `nullable`, so ✅.

On Android it looks like it does no get used at all, so ✅.

Will remove then.

hffmnn commented on packages/firebase_crashlytics/lib/src/firebase_crashlytics.dart:
```diff
@@ -56,6 +56,34 @@ class Crashlytics {
     }
   }
 
+  Future<void> onRuntimeException(dynamic exception, dynamic stack) async {
```
Naming is hard. I guess there are better names for this method?

hffmnn commented:
I signed it!

hffmnn commented:
> There's a bit of code duplication here, I think we should try to get more code re-use

Sure. Let me know if you would like to keep the 2 methods or funnel all through one.

> I think I'd like to rename onError to recordError to match other platforms, which will be a breaking change. This is a good opportunity to revisit the method signature as well.

I would start working on that, as soon as the API (one method vs two methods) decision is done.

collinjackson commented:
We've made a (tentative) decision to go with two methods:

```
recordFlutterError(FlutterErrorDetails details): Future<void>;
recordError(dynamic error, StackTrace stackTrace, { dynamic context }): Future<void>;
```

Example of usage:

```
void main() {
 FlutterError.onError = Crashlytics.instance.recordFlutterError;
 runZoned<Future<void>>(() async {
   runApp(MyApp());
 }, onError: Crashlytics.instance.recordError);
}
```

Let me know if you have any feedback, and if it looks good, I'd appreciate your help updating the PR. This is a breaking change, and we'll be making some other breaking changes at the same time, but I can handle that part.

hffmnn commented on packages/firebase_crashlytics/lib/src/firebase_crashlytics.dart:
```diff
@@ -24,35 +24,21 @@ class Crashlytics {
-        'keys': _prepareKeys(),
-      });
-    }
+  /// Submits non-fatal crash report to Firebase Crashlytics.
```
I think we might work on this doc comment.

collinjackson replied:
```suggestion
  /// Submits a report of a non-fatal error.
  ///
  /// For errors generated by the Flutter framework, use [recordFlutterError] instead.
```

hffmnn commented on packages/firebase_crashlytics/pubspec.yaml:
```diff
@@ -1,7 +1,7 @@
 description: Flutter plugin for Firebase Crashlytics. It reports uncaught errors to the
   Firebase console.
-version: 0.0.4+12
+version: 0.0.4+13
```
I guess the version has to be bumped to `0.5` to signal the breaking change?

collinjackson replied:
It should be 0.0.5

hffmnn replied:
Done

hffmnn commented:
@collinjackson I updated the PR with the new API and refactored the duplicated code. Let me know what you think.

Before I started I rebased to the latest `master` so I have no idea why some checks are failing with `Execution failed for task ':app:buildPluginDebugFirebaseAuth'.`.

collinjackson commented:
> @collinjackson I updated the PR with the new API and refactored the duplicated code. Let me know what you think.
> 
> Before I started I rebased to the latest `master` so I have no idea why some checks are failing with `Execution failed for task ':app:buildPluginDebugFirebaseAuth'.`.

Don't worry about these failures, they're affecting master as well https://cirrus-ci.com/task/4801326419017728

collinjackson commented on packages/firebase_crashlytics/example/lib/main.dart:
```diff
@@ -11,9 +13,14 @@ void main() {
 
   // Pass all uncaught errors to Crashlytics.
   FlutterError.onError = (FlutterErrorDetails details) {
```
Might be simpler here to just say `FlutterError.onError = Crashlytics.instance.recordFlutterError;`

collinjackson commented on packages/firebase_crashlytics/example/lib/main.dart:
```diff
@@ -11,9 +13,14 @@ void main() {
+
+  runZoned<Future<void>>(() async {
+    runApp(MyApp());
+  }, onError: (dynamic error, dynamic stack) async {
```
Might be simpler here to just say `onError: Crashlytics.instance.recordError`

collinjackson commented on packages/firebase_crashlytics/lib/src/firebase_crashlytics.dart:
```diff
@@ -24,35 +24,21 @@ class Crashlytics {
+  /// Submits non-fatal flutter crash report based on `FlutterErrorDetails`
+  /// to Firebase Crashlytics.
+  Future<void> recordFlutterError(FlutterErrorDetails details) async {
+    print('Error caught by Crashlytics plugin <recordFlutterError>:');
```
```suggestion
    print('Flutter error caught by Crashlytics plugin:');
```

collinjackson commented on packages/firebase_crashlytics/lib/src/firebase_crashlytics.dart:
```diff
@@ -24,35 +24,21 @@ class Crashlytics {
-  /// Submits non-fatal crash report to Firebase Crashlytics.
-  Future<void> onError(FlutterErrorDetails details) async {
-    print('Error caught by Crashlytics plugin:');
+  /// Submits non-fatal flutter crash report based on `FlutterErrorDetails`
```
```suggestion
  /// Submits report of a non-fatal error caught by the Flutter framework.
```

collinjackson commented on packages/firebase_crashlytics/CHANGELOG.md:
```diff
@@ -1,3 +1,7 @@
+## 0.0.4+13
+
+* Support `runZoned`s `onError` arguments to be send to Crashlytics.
```
We should describe the breaking change here.
```suggestion
* **Breaking Change** Renamed `onError` to `reportFlutterError`.
* Added `recordError` method for errors caught using `runZoned`'s `onError`.
```

collinjackson commented on packages/firebase_crashlytics/CHANGELOG.md:
```diff
@@ -1,3 +1,7 @@
+## 0.0.4+13
```
```suggestion
## 0.1.0
```

collinjackson commented on packages/firebase_crashlytics/example/lib/main.dart:
```diff
@@ -61,6 +68,17 @@ class _MyAppState extends State<MyApp> {
+              FlatButton(
+                  child: const Text('Async out of bounds'),
+                  onPressed: () {
+                    // Example of an exception that does not get catched
```
```suggestion
                    // Example of an exception that does not get caught
```

collinjackson commented on packages/firebase_crashlytics/example/lib/main.dart:
```diff
@@ -61,6 +68,17 @@ class _MyAppState extends State<MyApp> {
+                  child: const Text('Async out of bounds'),
+                  onPressed: () {
+                    // Example of an exception that does not get catched
+                    // by `Flutter.onError` but the `onError` handler of
```
```suggestion
                    // by `FlutterError.onError` but is caught by the `onError` handler of
```

collinjackson commented:
Looks good. Thanks for the contribution.
